[PowerPC] 32-bit large code-model support for toc-data

syzaara opened the pull request:
This patch adds the pseudo op ADDItocL for 32-bit large code-model support for toc-data.

amy-kwan commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -1236,6 +1249,30 @@ void PPCAsmPrinter::emitInstruction(const MachineInstr *MI) {
+
+    LLVM_DEBUG(assert(
+        !(MO.isGlobal() && Subtarget->isGVIndirectSymbol(MO.getGlobal())) &&
+        "Interposable definitions must use indirect access."));
```
nit: 
```suggestion
        "Interposable definitions must use indirect accesses."));
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:
```diff
@@ -6125,23 +6125,45 @@ void PPCDAGToDAGISel::Select(SDNode *N) {
+    // in the TOC entry, so we use an ADDItocL similar to the medium code
+    // model ELF abi.
+    if (isAIXABI &&
+        hasTocDataAttr(GA, CurDAG->getDataLayout().getPointerSize())) {
```
Maybe I am mistaken, but it doesn't look like `hasTocDataAttr()` uses `CurDAG->getDataLayout().getPointerSize()`. Do we need the second parameter?

Commit:
[PowerPC] 32-bit large code-model support for toc-data

This patch adds the pseudo op ADDItocL for 32-bit large code-model
support for toc-data.

chenzheng1030 commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -1147,15 +1147,28 @@ void PPCAsmPrinter::emitInstruction(const MachineInstr *MI) {
+      return GV->hasAttribute("toc-data");
+    };
+
+    if (!isSymbolTD(MO))
```
format related:
since the lambda function `isSymbolTD` is only used once, can we use anonymous lambda function in the if body? Like:
```
if ([](const MachineOperand &MO){}(MO)) {
}
```

Commit:
Address review comments

syzaara commented on llvm/lib/Target/PowerPC/PPCInstrInfo.td:
```diff
@@ -3346,11 +3346,13 @@ def ADDIStocHA : PPCEmitTimePseudo<(outs gprc:$rD), (ins gprc_nor0:$reg, tocentr
                    "#ADDItoc",
                    [(set i32:$rD,
                      (PPCtoc_entry tglobaladdr:$disp, i32:$reg))]>;
+def ADDItocL : PPCEmitTimePseudo<(outs gprc:$rD), (ins gprc_nor0:$reg, tocentry32:$disp),
```
I've added it to P10InstrResources.td and PPCMacroFusion.def.
getForwardingDefMI seems more involved and requires some investigation, so I'm thinking to leave that for a follow up patch.

chenzheng1030 commented on llvm/lib/Target/PowerPC/PPCISelDAGToDAG.cpp:
```diff
@@ -6144,23 +6142,44 @@ void PPCDAGToDAGISel::Select(SDNode *N) {
+                                            SDValue(Tmp, 0), GA));
+      return;
+    }
+
```
At line 6195, for 32-bit, should we use `ADDItocL` instead of `ADDItocL8`? Although seems there is no 32-bit case reaching that logic.(It uses `MVT::i64` at this phase, on 32-bit it should fail?)

syzaara replied:
Line 6195 is handling the Linux medium code model path. For tocdata 64-bit with large code model, I will have a follow up patch to use ADDItocL8 (with the appropriate enhancements made in PPCAsmPrinter.cpp). ADDItocL8 will then be used on line 6175 above.
I plan to post the 64-bit support next once this 32-bit patch is approved and merged.

chenzheng1030 replied:
OK. Thanks for explanation.

chenzheng1030 commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -1272,25 +1284,33 @@ void PPCAsmPrinter::emitInstruction(const MachineInstr *MI) {
+    if (Op == PPC::ADDItocL8)
+      assert((MO.isGlobal() || MO.isCPI()) && "Invalid operand for ADDItocL8.");
+    else
+      assert(MO.isGlobal() && "Invalid operand for ADDItocL.");
```
This looks weird to me. If `ADDItocL8` can be used for constant pool entries at 64-bit, why `ADDItocL` can not be used for contant pool entries at 32-bit?

syzaara replied:
ADDItocL is added specifically for toc-data on AIX. ADDItocL8 is already used on Linux medium code model for TOC access, so its possible that the MO is a CPI.  
ADDItocL however is only for toc-data symbols which must be globals.

chenzheng1030 replied:
> ADDItocL8 is already used on Linux medium code model for TOC access, so its possible that the MO is a CPI.

ADDItocL8 is used with CPI MO in global-isel when lowering a constant pool entry. But that is only for PPC64LE.

I originally thought ADDItocL should be used there for 32 bit. So maybe it is not strange that we use `ADDItocL` for CPI MO too. IIUC in the patch https://github.com/llvm/llvm-project/commit/34627e3434dbc8979218708ba4c4cf72f7c8e988 where `ADDItocL` was firstly introduced, there is no CPI MO usage for it.
But seems all 32-bit ABI(ELF32 and AIX32) does not allow direct access to TOC, so we should not use ADDItocL for 32bit, we should use LWZtocL.

I agree with you. We can limit ADDItocL to global address.

chenzheng1030 replied:
Code format:
```
assert((Op == PPC::ADDItocL8) ? (MO.isGlobal() || MO.isCPI()) : MO.isGlobal() && "Invalid operand for ADDItocL8.");
```

Commit:
Fix comment

chenzheng1030 reviewed (approved):
LGTM. I don't have further comments except one nit.

Commit:
Address review
